torch/csrc/autograd/variable.h

```diff
@@ -278,6 +285,54 @@ struct TORCH_API Variable : public at::Tensor {
   Impl* get() const;
 };
 
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+//                            Variable::AutogradMeta
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+struct TORCH_API Variable::AutogradMeta {
+  std::string name;
+
+  Variable grad_;
+  std::shared_ptr<Function> grad_fn_;
+  std::weak_ptr<Function> grad_accumulator_;
+
+  VariableVersion version_counter_;
+  std::vector<std::shared_ptr<FunctionPreHook>> hooks_;
+
+  // Only meaningful on leaf variables (must be false otherwise)
+  bool requires_grad_;
+
+  bool is_view_;
+
+  // The "output number" of this variable; e.g., if this variable
+  // was the second output of a function, then output_nr == 1.
+  // We use this to make sure we can setup the backwards trace
+  // correctly when this variable is passed to another function.
+  uint32_t output_nr_;
+  PyObject* pyobj_ = nullptr; // weak reference
+
+  // Mutex to ensure that concurrent read operations that modify internal
+  // state are still thread-safe. Used by get_grad_fn and
+  // get_grad_accumulator.
+  std::mutex mutex_;
+
+  virtual ~AutogradMeta() {}
+};
+
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+//                        Variable::DifferentiableViewMeta
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+struct TORCH_API Variable::DifferentiableViewMeta : public Variable::AutogradMeta {
```

Comment:
do we actually need this as part of this PR?  You are still going to have a Variable::Impl and a View::Impl for now, right?  So don't strictly need this yet?

Reply:
after this PR, `Variable::Impl` and `View::Impl` will become stateless operators that operate on `AutogradMeta` and `ViewMeta`. And since `View::Impl` assumes `AutogradMeta` to contain the `ViewMeta`-specific fields such as `base_` and `attr_version`, we likely need to have a `ViewMeta` that subclasses `AutogradMeta` and adds those fields in order to make it work. This new structure is also how the end state for `AutogradMeta` and `ViewMeta` would look like.